zppy/__main__.py

```diff
@@ -73,47 +73,54 @@ def main():  # noqa: C901
             raise OSError("Cannot create script directory")
         pass
 
+    if config["default"]["machine"] == "":
+        # MachineInfo below will then call `discover_machine()`,
+        # which only works on log-in nodes.
+        machine = None
+    else:
+        # If `machine` is set, then MachineInfo can bypass the
+        # `discover_machine()` function.
+        machine = config["default"]["machine"]
+    machine_info = MachineInfo(machine=machine)
+    default_machine = machine_info.machine
+    (
+        default_account,
+        default_partition,
+        default_constraint,
+        _,
+    ) = machine_info.get_account_defaults()
+    unified_base = machine_info.config.get("e3sm_unified", "base_path")
+
     # Determine machine to decide which header files to use
-    tmp = os.getenv("HOSTNAME")
-    if tmp:
-        machine_info = MachineInfo()
-        unified_base = machine_info.config.get("e3sm_unified", "base_path")
-        if tmp.startswith("compy"):
-            machine = "compy"
-            environment_commands = (
-                f"source {unified_base}/load_latest_e3sm_unified_compy.sh"
-            )
-            account = "e3sm"
-        elif tmp.startswith("cori"):
-            machine = "cori"
-            partition = config["default"]["partition"]
-            if partition not in ["haswell", "knl"]:
-                raise ValueError(
-                    f'Expected Cori partition to be "haswell" or '
-                    f'"knl" but got: {partition}'
-                )
-            # This will be the [default] environment_commands
-            environment_commands = (
-                f"source {unified_base}/load_latest_e3sm_unified_cori-{partition}.sh"
-            )
-            account = "e3sm"
-        elif tmp.startswith("blues"):
-            machine = "anvil"
-            environment_commands = (
-                f"source {unified_base}/load_latest_e3sm_unified_anvil.sh"
-            )
-            account = "condo"
-        elif tmp.startswith("chr"):
-            machine = "chrysalis"
-            environment_commands = (
-                f"source {unified_base}/load_latest_e3sm_unified_chrysalis.sh"
+    if config["default"]["machine"] == "":
+        config["default"]["machine"] = default_machine
```

Comment:
This will be `cori-haswell` or `cori-knl` instead of just `cori`, as zppy is expecting.  In the future, this will be `pm-cpu` and `pm-gpu` for perlmutter.

Would it be possible to adopt the `mache` convention here? There might be backwards compatibility issues if you do.  But you might be fighting against `mache` if you don't.

Reply:
Preparing for Perlmutter
--
I think since Cori is going to be shut down and replaced by Perlmutter, it doesn't make sense to spend time on cleaning up the Cori-specific code. (In the below section, I've listed some thoughts on what that would entail.)

I think for this PR, I can just change this line to use `cori` if `default_machine` is `cori-haswell` or `cori-knl`. Then, once `mache` has Perlmutter support, I can implement #310. I figure we can just delete all the Cori-specific code and replace it with more properly structured Perlmutter-specific code. Now, that would definitely be a breaking change, but I would argue since Cori won't be usable anymore, it won't really matter. (A user wouldn't be able to run a `machine = cori` config file anyway.)

Fixing Cori-specific code
--
This probably will be a breaking change (`zppy`'s major version will be increased). That said, it may be possible to do some sort of automatic conversion if `zppy` receives parameters in today's expected form. In that case, we could avoid a breaking change. For example, if a user lists `cori` as `machine`, and `haswell` as `partition` (or `constraint` as you suggest in your other comment), we could infer `cori-haswell` is the machine to pass to `machine_info = MachineInfo(machine=machine)`.

```
$ cd zppy/zppy # On the branch for this PR
$ git grep -n "cori"
__main__.py:102:        if config["default"]["machine"] == "cori":
__main__.py:106:    if config["default"]["machine"] == "cori":
__main__.py:115:        if config["default"]["machine"] == "cori":
__main__.py:118:            ] = f"source {unified_base}/load_latest_e3sm_unified_cori-{partition}.sh"
templates/amwg.csh:439:{% elif machine == 'cori' %}
templates/amwg.csh:3298:{% if machine == 'cori' %}
templates/amwg.csh:3299:# For NERSC cori, make sure it is world readable
templates/e3sm_diags.bash:486:{% if machine == 'cori' %}
templates/e3sm_diags.bash:487:# For NERSC cori, make sure it is world readable
templates/e3sm_diags.bash:508:{% if machine == 'cori' %}
templates/e3sm_diags.bash:509:# For NERSC cori, change permissions of new files
templates/global_time_series.bash:120:{% if machine == 'cori' %}
templates/global_time_series.bash:121:# For NERSC cori, make sure it is world readable
templates/global_time_series.bash:142:{% if machine == 'cori' %}
templates/global_time_series.bash:143:# For NERSC cori, change permissions of new files
templates/ilamb.bash:93:{% if machine == 'cori' %}
templates/ilamb.bash:94:# For NERSC cori, make sure it is world readable
templates/ilamb.bash:116:{% if machine == 'cori' %}
templates/ilamb.bash:117:# For NERSC cori, change permissions of new files
templates/mpas_analysis.bash:127:{% if machine == 'cori' and partition == 'knl' %}
templates/mpas_analysis.bash:130:{% elif machine in ['cori', 'compy', 'anvil', 'chrysalis'] %}
templates/mpas_analysis.bash:137:{% if machine == 'cori' and partition == 'knl' %}
templates/mpas_analysis.bash:140:{% elif machine in ['cori', 'anvil', 'chrysalis'] %}
templates/mpas_analysis.bash:148:{% if machine == 'cori' and partition == 'knl' %}
templates/mpas_analysis.bash:151:{% elif machine in ['cori'] %}
templates/mpas_analysis.bash:172:{% elif machine == 'cori' %}
templates/mpas_analysis.bash:344:{% if machine == 'cori' %}
templates/mpas_analysis.bash:345:# For NERSC cori, make sure it is world readable
templates/mpas_analysis.bash:365:{% if machine == 'cori' %}
templates/mpas_analysis.bash:366:# For NERSC cori, change permissions of new files
templates/slurm_header.sh:14:{% elif machine == 'cori' %}
templates/slurm_header.sh:16:# Running on cori-haswell or cori-knl depending on 'partition'
templates/tc_analysis.bash:81:{% if machine == 'cori' %}
templates/tc_analysis.bash:98:{% if machine == 'cori' %}
templates/tc_analysis.bash:104:{% if machine == 'cori' %}
```
We'd have to change all those `if` `jinja` statements to check for both `cori-haswell` and `cori-knl` instead.

Reply:
Good point, too much to take on for a machine that's about to disappear.